### ISSUE TITLE
Istio-pilot crash due to high memory leak 

### ISSUE THREAD

**Issue by ShimiT** [istio/istio#14782]
**Bug description**

We installed istio via helm default config (prod) and try to simulate a high load environment according to our production KPI's.

Each pod uses `8 route`, each pod created in new ns.

Using the default configs `pilot` was **crashed** after the creation of approximately `800` pods. 
To overcome this we used helm [config](https://istio.io/docs/reference/config/installation-options/#pilot-options) to extend the default memory to `4096MB` and `autoscaleMax` to `15`,
with this config we were able to create `1250` pods until we got the following error from istio-pilot. (OOM)

Error:
`The node was low on resource: memory. container discovery was using 12687204ki, which exceeds its request of 4Gi` 
( before we change the config we got the same error with lower numbers of memory etc)

Clearly, we tackled a memory leak as pilot reach the memory limit of `60GB`  after a 10-15 min as we are creating new pod per one second. 
we also tried to reduce the load and create a like following:
1 every 12 seconds -> 5 in a minute > 300 in a hour -> 3000 in 10 hours.
Again, after ~1,300 pods creation, all `istio-pilot` active connections **crushed**.

Following is some snapshots: 
( IMO the graph is not accurate as pilot memory show more than 100GB of memory usage…)  

`pilot`
https://snapshot.raintank.io/dashboard/snapshot/BHHzAQQU6eKCSwenvAKWry4hh2QSbS0M?orgId=2

https://snapshot.raintank.io/dashboard/snapshot/sZqEgXIRCJ46lfvkreVp68RbZS0ZL50v?orgId=2&from=1560152601034&to=1560153201034

`performance`
https://snapshot.raintank.io/dashboard/snapshot/OQau3FszAH1tcR6MBOu0n7PM33e0uiLE?orgId=2

**Affected product area (please put an X in all that apply)**
[ ] Configuration Infrastructure
[ ] Docs
[x] Installation
[x] Networking
[x] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience

**Expected behavior**
istio pods should be functional 
**Steps to reproduce the bug**

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
we are using K8S version `1.13.6` and istio `1.1.7`
**How was Istio installed?**
`First run: `

`helm template install/kubernetes/helm/istio --name istio --namespace istio-system | kubectl apply -f -`

`Second run: ` scaling istio-pilot & ingress-gateway

`helm template install/kubernetes/helm/istio --name istio --namespace istio-system \
     --set pilot.autoscaleMax=15 --set pilot.resources.requests.memory=4096Mi \
     --set gateways.istio-ingressgateway.autoscaleMax=15 --set global.proxy.resources.limits.memory=256Mi \
     --set tracing.enabled=true --set servicegraph.enabled=true \
     --set grafana.enabled=true | kubectl apply -f -` 

**Environment where the bug was observed (cloud vendor, OS, etc)**
K8S 

in addition, We didn't use the label-inject (side-car) option for each pod yet.

**Comment by ShimiT** [comment]
If some data is missing please let me know as currently we are blocked by `pilot` and cannot use istio.

**Comment by howardjohn** [comment]
Memory is probably double counted due to https://github.com/istio/istio/issues/14705. 

Good to know its without injected pods, that isolates the problem space.

So just from having 1.4k services in its registry, each pilot is using ~5 CPU and ~10GB of memory (probably half those numbers due to dashboard bug). Definitely higher than expected; I just created 1000 ServiceEntries and pilot used 50mb of memory.

Can you capture a heap profile and CPU profile? I wrote some instructions on a similar issue https://github.com/istio/istio/issues/14537 (should add this to the wiki probably)

**Comment by howardjohn** [comment]
From some profiles sent to me:

`route.CombineVHostRoutes` usage is crazy high. 33% of allocations were from there.
`DefaultSidecarScopeForNamespace` mostly in the VirtualService part also uses another 27%, so total of 60% of memory on virtual service related stuff.

So do you have one VS for each pod, so ~1250 of the VS with 8 routes?

**Comment by howardjohn** [comment]
ok created 1000 VS. Each gateway is using 5-600Mb, pilot is using 2gb. So a a lot but not as much as you see. I don't see `route.CombineVHostRoutes` or `DefaultSidecarScopeForNamespace` using any substantial amount

**Comment by ShimiT** [comment]
@howardjohn  I've used the same test on our cluster, i.e. `1000VS`   with `8000` routes **without pods**  and `istio-pilot` works as expected. `pilot` uses only one replica with 1-2GB at load. 
Our scenario is a bit different,  `1000 namespaces-> 1000  pods-> 8000 routes` which pilot cannot handle even with `15` instances (4GB memory each). if you want me to provide additional info please let me know.

**Comment by RaynDol** [comment]
HI @howardjohn @ShimiT , are you able to made some progress , we are struggling with the similar problem (pilot) almost 3 weeks without success , we wanted to switch to `istio` for production and **pilot is currently blocker** for us as it consume `80GB` (we try to scale it to 10 replicas with 8GB memory each) of memory before it crushed , please advice.

**Comment by ShimiT** [comment]
@howardjohn  - How we proceed with this issue?

**Comment by jkbschmid** [comment]
@ShimiT
I'm trying to reproduce your setup, but I somehow cannot reproduce the error.
Would you mind providing some more details on what you exactly deployed and in which namespace? Do the pods you deploy have a sidecar or not?
I tried creating 1000 namespaces that each contain 1 Service, 1 Pod (no sidecars) and 1 VirtualService, but memory usage of pilot was normal.

**Comment by jsPlachetka** [comment]
We took a deeper look into the memory consumption while creating ~ 1000 namespaces and virtualservices and came up with a potential fix.

@ShimiT could you try your scenario again using this image `gcr.io/peripli/pilot:issues-14782` and tell us if this improved the memory consumption?

**Comment by howardjohn** [comment]
@jsPlachetka @kramerul can you post the config you are using somewhere so I can easily reproduce what you have?

**Comment by ShimiT** [comment]
we are currently running the tests, and It seems that the new pilot image fixes the memory leak issue, I'll update when we finish all our test cycle.

**Comment by kramerul** [comment]
> @jsPlachetka @kramerul can you post the config you are using somewhere so I can easily reproduce what you have?

[14782.tar.gz](https://github.com/istio/istio/files/3430091/14782.tar.gz)

**Comment by ShimiT** [comment]
@howardjohn - After further validations, it seems that the new image gcr.io/peripli/pilot:issues-14782 fixes the memory leak in Pilot. is it possible to down port the fix to v1.2.3?

**Comment by edussx** [comment]
I cherry picked the fix provided by @jsPlachetka to release 1.1.11 and built my own image. Now I see the memory footprint is pretty normal and no more OOM kill happened in the recent 12 hrs.

I'm seeing if https://github.com/istio/istio/issues/16094 will still happen without OOM-killed Pilot

**Comment by howardjohn** [comment]
With PR https://github.com/istio/istio/pull/16211 this is greatly improved. Basically the problem was inefficiencies specifically related to having a low (less than 100) number of connections to pilot. In all the posts here there are only a few connections to pilot (gateways) but lots of config.

https://github.com/istio/istio/pull/15810 takes a different approach, ensure that only one push context is active a time. This is needed because pilot uses a TON of CPU due to issues fixed in #16211, making the PushContext calculation take a really long time, just because pilot has its CPU throttled. Once pilot has the CPU throttled, the whole problem cascades because it is creating a ton of allocations, but never finishes using them due to the CPU throttle.

We can probably merge both changes and this problem will be completely gone

**Comment by edussx** [comment]
@howardjohn I didn't quite understand why `having a low (less than 100) number of connections to pilot` causes pushing slow(inefficiencies). Could you explain a little bit more?

**Comment by howardjohn** [comment]
Sure. It is purely an implementation detail, but what happens is there is a queue that stores all proxies that need an update. This will deduplicate repeated updates to proxies. As we dequeue these, we send them on an unbuffered channel, meaning it will block, preventing multiple concurrent pushes to the same proxy. 

The code has a limit of only dequeueing 100 proxies at a time. This means if you have 100+ proxies, we are unlikely to be stuck trying to send on the channel, because the proxy we dequeue will be unlikely to be currently pushing.

If we only had 1 connection, then what happens is we will hve 100 changes all stacked up trying to send on the channel, rather than deduping in the queue.

So basically if we have few proxies then we don't dedupe pushes correctly -- except we will now in 1.3

**Comment by c0d1ngm0nk3y** [comment]
@howardjohn We compared the results of #16211 and #15810 . Although they sound like they are tackling the same problem, the memory consumption in our scenario is quite different. The memory usage still grows exponential and not linear. So we think #15810 would still be important. Do you think both approaches can be combined?

Worth to mention maybe that we do not use sidecars at all in our scenario. So there is only one proxy (istio-ingress).

![image](https://user-images.githubusercontent.com/13679746/63007922-eabfae80-be81-11e9-958c-397a909210c3.png)

**Comment by edussx** [comment]
@howardjohn Thanks for the explanation! I suppose you're talking about the queue implemented here: https://github.com/istio/istio/pull/15405/files#diff-5c8e05caf018e8f8a83c7e14133f7627R44

But both @ShimiT 's and my cases happened in release 1.1.x which does not have the queue introduced. (Yes I didn't find the code because I was looking at 1.1.x at that time :) )

I notice there're a bunch of PRs besides https://github.com/istio/istio/pull/16211 in release-1.3 branch addressing this performance issue so probably I need upgrade our Istio and then rebuilt Pilot on the release-1.3 branch.

**Comment by howardjohn** [comment]
@c0d1ngm0nk3y yes I don't think the PR I linked means we don't want the other one, I submitted it for other reasons then noticed it made an improvement here. I think the one you have helps in overloaded scenarios - when pilot is throttled, it has no backoff and keeps creating more and more objects that it can't free fast enough.

**Comment by howardjohn** [comment]
Changes are in, I have verified myself that this is much improved - if someone else could verify as well that would be great. Thanks for all the help, debugging, etc on this issue!

**Comment by c0d1ngm0nk3y** [comment]
We verified it with the same test as before. With v1.3-rc0, the error occured. With v1.3-rc1, the memory consumption looks good (as expected from @howardjohn)

![image](https://user-images.githubusercontent.com/13679746/64167790-6cd04100-ce4a-11e9-806d-4aa6342e056b.png)
